### PR TITLE
Implemented  getAllFlyerCategories and tests

### PR DESCRIPTION
**Pull request by cindy-x-liang**
## Overview

Created getAllFlyerCategories which will return all the different flyer categories of the flyers inside the database.

## Changes Made

Create an async function called getAllFlyerCategories in FlyerRepo.ts. Used the mongoose function distinct() to get unique category slugs from each flyer. Created a query for this function in FlyerResolver.ts, and implemented test cases in flyer test.t

## Test Coverage

Test cases test for instances when all flyers have different categories and when flyers share the same category.

## Screenshots (delete if not applicable)

<details>

  <summary>GraphQL Screenshot</summary>

Screenshot 2023-09-27 at 6 00 21 PM

  

</details>

### CODE REVIEW

File: src/repos/FlyerRepo.ts
Comment (zachseidner1) [inline src/repos/FlyerRepo.ts:82]: The `.then` method is not needed here. When the `FlyerModel.collection.distinct` call finishes, the `.then` returns a new promise, whose behavior is defined based on the function you put in. But your function just returns the resulting value, so it is redundant. See [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then#return_value) for more info.

**Review by zachseidner1 — CHANGES_REQUESTED** [review]
Nice job! Just see my comment about the `.then` statement.

File: src/repos/FlyerRepo.ts
Comment (vinnie4k) [inline src/repos/FlyerRepo.ts:82]: Do you have Prettier installed? ESLint should also point out formatting errors. Replace this function with:
```
/**
 * Looks for unique category slugs for all the Flyers
 *
 * @returns a list of Strings with all the category names.
 */
const getAllFlyerCategories = async (): Promise<string[]> => {
  return FlyerModel.collection.distinct('categorySlug').then((categories) => {
    return categories;
  });
};
```
Then, address @zachseidner1 's comment on removing `.then`.

File: src/repos/FlyerRepo.ts
Comment (vinnie4k) [inline src/repos/FlyerRepo.ts:284]: This should be sorted alphabetically.

File: src/resolvers/FlyerResolver.ts
Comment (vinnie4k) [inline src/resolvers/FlyerResolver.ts:161]: Same as above, replace this line with:
```
@Query((_return) => [String], {
  nullable: false,
  description: `Returns a list of <Strings> representing all of the categories of the different flyers in this database`,
})
async getAllFlyerCategories() {
  return FlyerRepo.getAllFlyerCategories();
}
```

File: src/resolvers/FlyerResolver.ts
Comment (vinnie4k) [inline src/resolvers/FlyerResolver.ts:163]: Remove this line.

File: src/tests/flyer.test.ts
Comment (vinnie4k) [inline src/tests/flyer.test.ts:315]: Remove this line.

File: src/tests/flyer.test.ts
Comment (vinnie4k) [inline src/tests/flyer.test.ts:279]: I would put all of this at the end of the file just to keep things in order of implementation but not a big deal.

**Review by vinnie4k — CHANGES_REQUESTED** [review]
Great work! I am very impressed in your ability to see the common patterns in our codebase and being able to learn from them in a short amount of time, especially since you don't have a background in TypeScript.

Please address my comments about formatting and address @zachseidner1 's comments. Then request me again. I would also wait for @Archit404Error the goat to review before merging.

File: src/repos/FlyerRepo.ts
Comment (Archit404Error) [inline src/repos/FlyerRepo.ts:77]: This function actually doesn't need to be `async`, since all you're doing here is returning a promise, which should actually be a synchronous action. Async/await stuff can be kind of confusing -- as a general rule of thumb, only mark a function `async` if you're using the keyword `await` inside of it!

File: src/tests/flyer.test.ts
Comment (Archit404Error) [inline src/tests/flyer.test.ts:291]: A more idiomatic way to write this code would be use an array and leverage `createSpecfic()`'s ability to create more than one of a type of flyer, especially since `insertMany()` allows you to insert multiple documents at once:
```suggestion
    const flyerSlugs = ["Academic", "Dance"];
    let generatedFlyers = [];
    generatedFlyers.concat(await FlyerFactory.createSpecific(1, {categorySlug: flyerSlugs[0]}))
    generatedFlyers.concat(await FlyerFactory.createSpecific(2, {categorySlug: flyerSlugs[1]}))
    
    await FlyerModel.insertMany(generatedFlyers)
```

The code is now a lot shorter/simpler with less variable names to keep track of, while still giving you the same functionality

File: src/tests/flyer.test.ts
Comment (Archit404Error) [inline src/tests/flyer.test.ts:309]: I'd suggest rewriting this in the same way I suggested for the test above

File: src/tests/flyer.test.ts
Comment (Archit404Error) [inline src/tests/flyer.test.ts:280]: Our goal should be to have really descriptive test names -- I know a lot of the other tests in this codebase are structured like the name you've written here, but I'd suggest changing it up to reflect what we're actually testing. Our goal here is to show that our query filters out duplicate slugs, so something like `Handle non-unique CategorySlugs across flyers in collection` or `ensure only unique categorySlugs are returned` would make more sense here

**Review by Archit404Error — CHANGES_REQUESTED** [review]
Great job on your first PR! I think Vin and Zach covered a lot of the formatting details -- once you get your ESLint and auto formatting going you should be able to resolve those quickly.

I wanted to highlight some areas of your code that could be optimized a bit more for readability -- the goal is always to make sure we have idiomatic code that is easy to understand and modify, so you can check those comments out below

Overall, though, great job on this PR!

File: src/tests/flyer.test.ts
Comment (vinnie4k) [inline src/tests/flyer.test.ts:298]: Keep the text inside of `describe` the same as you had before: `describe('getAllFlyerCategories tests', () => {`
What Archit meant was for the `test` part since each test case is testing for something different. At some point, we will need to refactor the entire testing suite and write test cases for resolvers.

**Review by vinnie4k — APPROVED** [review]
Fix your test description and you should be good to go on my part. Once @zachseidner1 and @Archit404Error approve, feel free to merge. Also, make sure you mark the comment with "Resolve conversation" so that we know you have resolved it. Great work!

File: src/tests/flyer.test.ts
Comment (zachseidner1) [inline src/tests/flyer.test.ts:326]: Don't repeat yourself! Replace this with a for loop:
```TypeScript
for (const slug of flyerSlugs) {
      generatedFlyers.push((await FlyerFactory.createSpecific(1, { categorySlug: slug }))[0]);
    }
```

File: src/tests/flyer.test.ts
Comment (zachseidner1) [inline src/tests/flyer.test.ts:307]: You may want to make this a for loop as well so the tests follow the same style
```TypeScript
for (const slug of flyerSlugs) {
      generatedFlyers.push((await FlyerFactory.createSpecific(1, { categorySlug: slug }))[0]);
    }
```

**Review by zachseidner1 — CHANGES_REQUESTED** [review]
I know my comments tend to be critical, but I do want to emphasize that this is great work for your first PR, especially considering you don't have much experience with TypeScript. Just work a bit on code concision, but other than that awesome work!

File: src/tests/flyer.test.ts
Comment (Archit404Error) [reply src/tests/flyer.test.ts:307]: I actually wouldn't recommend using a for loop here because the goal is to create 3 flyers, two of which have the same slug. The code above would only test 2 flyers with 2 unique slugs which is semantically what the other test does. You could technically make this a loop by making the categorySlugs array ["Category1", "Category2", "Category2"] but honestly I think the current code is a lot more readable

**Review by Archit404Error — DISMISSED** [review]
Good to go once you resolve all comments above!

File: src/tests/flyer.test.ts
Comment (zachseidner1) [reply src/tests/flyer.test.ts:307]: Interesting, I would argue that the `["Category1", "Category2", "Category2"]` is more readable, since it clearly outlines which category slugs we are adding to `generatedFlyers` at the top of the function, as opposed to the current code, where one might skip over the `2` in the parameters of the `createSpecific` function (this is what I did while reading it). But we're just nitpicking at this point, it doesn't really matter which approach you use. That's why I said you _may_ want to make this a for loop. But imo it should be up to Cindy at this point.

Comment (zachseidner1) [reply src/tests/flyer.test.ts:307]: Wait just noticed that the current code actually only adds 2 distinct Flyers!
`(await FlyerFactory.createSpecific(2, { categorySlug: flyerSlugs[1] }))[0]`
only adds the Flyer at index 0 of the generated array of 2 Flyers. So she needs to either use the for-loop approach or use `generatedFlyers.concat` and remove the accessor.

Comment (Archit404Error) [reply src/tests/flyer.test.ts:307]: Oh yeah I had assumed this code was using `concat()` like my suggestion above haha.

**Review by zachseidner1 — DISMISSED** [review]

File: src/tests/flyer.test.ts
Comment (Archit404Error) [inline src/tests/flyer.test.ts:303]: Using concat is cleaner than trying to unpack the array:

```suggestion
      generatedFlyers.concat(await FlyerFactory.createSpecific(1, { categorySlug: slug }));
```

File: src/tests/flyer.test.ts
Comment (Archit404Error) [inline src/tests/flyer.test.ts:316]:
```suggestion
      generatedFlyers.concat(await FlyerFactory.createSpecific(1, { categorySlug: slug }));
```

**Review by Archit404Error — DISMISSED** [review]
Good to go once you clean up these two lines

**Review by Archit404Error — APPROVED** [review]
Good to go

**Review by vinnie4k — APPROVED** [review]
Good to go.

**Pull request merged**
